File: Maths_And_Stats/Series/Sum_of_AP/Sum_of_AP.js
/*
Arithmetic progression if a sequence of number has same common difference. 
Here
    a is first term of series
    d is common difference
    n is number of terms
*/

const readline = require('readline')

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

const inputParams = () => new Promise(resolve =>
    rl.question('\nEnter first term (a)\n> ', a =>
        rl.question('Enter common difference (d)\n> ', d =>
            rl.question('Enter the number of elements (n) \n> ', n => {
                a = parseInt(a)
                d = parseInt(d)
                n = parseInt(n)
                rl.close()
                resolve({ a, d, n })
            })
        )
    )
)

const sumOfAP = (a, d, n) => {
    let sum = 0;
    let i = 0;
    while (i < n) {
        sum += a;
        a += d;
        i++;
    }
    return sum;
}

inputParams()
    .then(({ a, d, n }) => {
        console.log(`\nSum of arithmetic progression: \
  ${sumOfAP(a, d, n)}`)
    })
/*
For testin purpose
let a = 1;
let d = 1.5;
let n = 10;

console.log(sumOfAP(a, d, n));
*/

File: Maths_And_Stats/Series/Sum_of_AP/Sum_Of_AP.py
# Function to calculate sum of a given AP sequence
def Sum_AP(ap):
    terms = len(ap) # Number of terms in AP
    common_difference = [] # A list for storing all the common differences
    index = 0
    # Loop to get all common differences
    while index < (terms-1):
        # adding the common difference of current and net element
        common_difference.append(ap[index+1]-ap[index])
        # if this and previous common difference are not same and if index is not zero, sequence is not in AP
        if (index!=0) and (common_difference[index] != common_difference[index-1]):
            return "Not in AP"
        index += 1
    # this part will be executed only if sequence is in AP

    # if no of terms is even
    if terms % 2 == 0:
        # if all common difference are same, index will be term - 1
        if index == (terms - 1):
            # implementing the formula for sum of AP with even no of terms= (n/2)*(2*a +(n-1)*d)
            sum = (terms // 2) * ((2 * ap[0]) + ((terms - 1) * common_difference[0]))
            return "Sum of AP = "+str(sum)

    # if no of terms is even
    if terms % 2 == 1:
        # if all common difference are same, index will be term - 1
        if index == (terms-1):
            # implementing the formula for sum of AP with odd no of terms = n*(a+((n-1)/2)*d)
            sum = terms*(ap[0]+((terms-1)//2)*common_difference[0])
            return "Sum of AP = "+str(sum)


# Driver Code

# Calling function Sum_AP and passing into it the sequence given as input
ap = list(map(int,input("Enter elements : ").split()))
print(Sum_AP(ap))

'''
Input description: 
    Line 1 -- Sequence (seperated by spaces)

Output Description :
    Sum of AP = Sum

Example Outputs:

    Input 1:
    Enter elements : 1 2 3 4
    
    Output 1:
    Sum of AP = 10
    
    
    
    Input 2:
    Enter elements : 1 3 5
    
    Output 2:
    Sum of AP = 9
'''